Accessibility Improvements

jasonhenriquez opened the pull request:
---
Accessibility Improvements
---

**Pull Request Type**
- [x] Feature Implementation

**Related issue**
closes #693 
closes #1580

**Description**
Currently, most of FreeTube's controls do not use the proper semantic elements (e.g., `button`, `a`). These "fake controls" (i.e., mostly `<div>`s, `<span>`s, and custom FreeTube elements with a `@click` or `@change` event handler) seem to work fine for sighted users, but they lack much of the functionality & screen reader interaction provided by the native control elements.

To minimize the effect on the look and feel of the app, this pull request uses a combination of roles (e.g., `button`, `link`), accessibility attributes (e.g., `aria-label`, `aria-selected`), and event handlers (e.g., `@keydown`) to enhance FreeTube's controls (e.g., `<ft-icon-button>`, `<ft-notification-banner>`) & assorted existing "fake controls" with the proper behavior. Almost all controls should now be both keyboard focusable and keyboard interactive.

The appearance of the app is virtually unaffected, but some elements have new styling on being focused (and hovered for some). For people who do not use screen readers or keyboard navigation, most of the focusing style changes should be insignificant. Sighted users will appreciate that some buttons, links, and tabs now appropriately change appearance on hover.

While most users do not use their keyboard to navigate the app, the user experience for all users should be noticeably improved. The profile selection tab, channel page tabs, and all drop-down lists can now be navigated with arrow keys. All video lists can now be navigated with arrow keys as well.
 
**Screenshots (if appropriate)**
The visual appearance of the app is the same. The main differences relate to interactivity.

**Testing (for code that is not small enough to be easily understandable)**
I am neither Blind nor competent with a screen reader. I did not use a screen reader in my testing, but I did thoroughly test keyboard navigation. Although these improvements should be unilaterally positive, I would be welcome to hear the input of people who regularly use screen readers / keyboard navigation.

Here are two videos testing/showcasing most of the new functionality in five separate clips. The only changes not displayed in the videos are comment navigation, live chat navigation, and new benefits for screen readers (e.g., new button labels): 

 - https://user-images.githubusercontent.com/84899178/125984326-abbbc140-9046-407f-81d5-6008d1893491.mp4

 - https://user-images.githubusercontent.com/84899178/125984380-67135c0e-e383-4909-a577-dab587cf120e.mp4


**Desktop (please complete the following information):**
 - OS: Windows
 - OS Version: 10
 - FreeTube version: 0.13.2 Beta

**Additional context**
This pull request replaces all the "fake controls" I could find. For "fake controls" beginning with "ft-", the appropriate `@keydown` behavior is now handled automatically. However, future developers should take care to still give the appropriate role [most likely either `role="button"` or `role="link"`] to their `<ft-channel-bubble>`s and `<ft-profile-bubble>`s upon creating them. If you are thinking about using a `<font-awesome-icon>` as a button or link, you should probably use an `<ft-icon-button>` instead. If you have no choice but to use a fake control, append to it `@keydown.enter.prevent="linkOrButtonFunction"` if it's a link or button as well as a `@keydown.space.prevent="buttonFunction"` if it's a button.

Future FreeTube contributors should take care to not use "fake controls" (e.g., a `<div>` or `<span>` with a `@click` or `@change` event handler). This is inaccessible to users who use keyboard navigation and/or screen readers to navigate the app. Whenever possible, use native controls (e.g., `<button>`, `<a>`) or FreeTube's controls (e.g., `<ft-button>`, `<ft-icon-button>`). Otherwise, you (or someone else) will have to do the work of adding roles, aria-labels, accessible event handlers, and so on. I would like these accessibility suggestions to be added to the wiki/documentation and/or ~~required by linter rules~~.

2 Final Updates:
1. 95d1380: Linter rules against inaccessible fake controls, among other accessibility rules, will be codified in the FreeTube linter upon the passage of this PR. This was done using all but two of the recommended rules in `eslint-plugin-vuejs-accessibility` (see discussion [here](https://github.com/vue-a11y/eslint-plugin-vuejs-accessibility/issues/54) and [here](https://github.com/vue-a11y/eslint-plugin-vuejs-accessibility/issues/97)). 
2. 3ad708e: This pull request has now implemented all of its intended accessibility changes and more. With the helpful input of @RastislavKish and four additional waves of changes since July, this PR has finally begun being tested.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
legendary achievement unlocked

RastislavKish commented:
This seems very interesting, I would like to try it out. I've cloned the repository and switched to the improveAccessibility branch (6d7d9755b57).\
I've navigated to the folder and used commands:\
```npm install```\
```npm run-script dev-runner```

The interface pops up, I can do a search and open a video, but it won't play, the duration shows zeroes (with the official AppImage working as usual).

Is this an expected behavior, or did I miss something? I apologize if this is a noob question, while I am a programmer, I don't have any experiences with compiling/running Electron apps from source nor with using node.js. :)

My environment: Ubuntu mate 20.04 64-bit, node.js v12.22.2, npm 6.14.13.

ChunkyProgrammer commented:
@RastislavKish try doing `npm run dev`

jasonhenriquez commented:
@RastislavKish Were you able to get it working? I suspect something was wrong with your configuration, as I cannot replicate these errors.

RastislavKish commented:
@jasonhenriquez yes, it turned out Node JS 12.x had problems with some dependencies, upgrading to 16.x resolved the issue.

First of all, I'd like to say this is really a nice job! I've tested the controls and not just buttons, but also menus (the video formats menu, video options menu etc.) work properly now.

I could not check the video comment section, as the button did not react, but from FreeTube Issues I guess this is rather a bug in the app than accessibility.

Currently, I have just one thing to note. Making elements accessible had an interesting sideeffect - lists of videos (like thatone on the search page) have about three elements between each item.

To explain what I mean, imagine you search for a video, and the only thing you see is the currently keyboard focused element. You press tab to look up the results, but instead of having to press it just one time to get from one item to another, you need to do it 4 times to get through:
* The link to author's channel
* The Save video button of the next result
* The Show options button of the next result

And just then comes the actual result. The same goes backward of course, just in reversed order, so detailed checking of the search results can turn into a quite lengthy process.

usually, search results are marked as headings (Google, Youtube, DuckDuckGo, Teddit), and screenreaders have shortcuts to jump to previous and next heading, so users can navigate quickly from item to item and deal with specific elements just on places where they need them.

So the question is, could they possibly be implemented here in a visually non-disturbing way, so the lists navigation would be more confortable?

Anyway, this is really a huge accessibility improvement, thanks for the great job!

jasonhenriquez commented:
@RastislavKish I was able to implement your suggested changes for all "search results" (including lists of videos, playlists, and channels), and the "Show Options" button is the only element before the heading in tab order. However, it is extremely difficult to move the heading before the "Show Options" button in tab order without affecting the appearance at all. The current appearance relies on the use of "float: right" on the "Show Options" button, which only works correctly if it is placed first before its siblings.

As it stands, users traversing headings with a screen reader will have to go "backwards" once to access the settings of a particular video. Is this an adequate solution to the issue you described @RastislavKish?

RastislavKish commented:
@jasonhenriquez yes, it's perfect!

As for the order of elements, I think it doesn't really matter, this is quite a frequent thing to look for, so we are used to check for it. :)

Moving even the Save button is technically not necessary, I think ab332854 is fine. The most important thing is the quick navigation, and that is achieved now, so the lists are confortable to navigate and use.

GilgusMaximus commented:
We'll have to check in with dragos for this. It looks really good and is something I planned to work on for a long time, so i am very happy someone did it. But I would imagine that we will probably merge this before redoing the whole UI with dragos over at #1503 depending on how well it works out.

Some testing is definitely required once we fixed the comments

RastislavKish commented:
@jasonhenriquez I have pulled your merge in order to test the comments section. It works really nice, there is just one minor issue.

Say there is a comment with 100 replies. When I press View 100 replies, just say 10 are displayed (I did not count the exact amount).\
That's fine, I read through the replies and assume they picque my interest, so I want to read more. After pressing the Show more replies button, more replies are loaded, but the focus stais on the button, so I need to arrow-up all the way to the point where I ended, so I can read in order.\
The same goes for the Load more comments button.

So, the question is, would it be possible to move the cursor to the first newly loaded reply / comment after pressing one of the load more buttons?

jasonhenriquez commented:
> So, the question is, would it be possible to move the cursor to the first newly loaded reply / comment after pressing one of the load more buttons?

This should work now as of the latest commit.

RastislavKish commented:
@jasonhenriquez yes, it's perfect!

I don't see any other accessibility issues, the interface is consistent and intuitive to use. Thanks again for all the great job, I hope we will see this in a release soon. :)

peepopoggers commented:
:+1:

RastislavKish commented on src/renderer/components/ft-channel-bubble/ft-channel-bubble.vue:
```diff
@@ -1,11 +1,15 @@
     <img
       class="bubble"
       :src="channelThumbnail"
+      alt=""
```
@jasonhenriquez which UI part is it? I can test if it's presented correctly.

jasonhenriquez replied:
Thank you so much for helping with this @RastislavKish! This control is a link to a given user's channel. One place it appears many times is on the side navigation bar as a list of channels that someone is subscribed to. It may also appear under the "About" section of a given channel as one of its related channels.

RastislavKish replied:
@jasonhenriquez so it should be read as the linked channel's name, right?
I can confirm this works correctly, I see all my subscribed channels listed on the sidebar, activating any of them brings me to the appropriate channel.

Feel free to ask if anything requires testing, I'm keeping my eye on this thread as well as the new commits, so I'm ready to try anytime!
Thanks for all the effort!

jasonhenriquez replied:
Good, it's working as intended! You have been a spectacular help @RastislavKish. The app will be more accessible to so many more people because of your testing and input. You do not have to do any further work on this if you don't want to. The only other new changes at this time to test are modal accessibility (e.g., when using "Import Subscriptions" on the setting page) and left & right arrow key navigation of videos.

peepopoggers commented:
Why did package.lock change?

jasonhenriquez commented:
> Why did package-lock.json change?

I had to update a dependency for `eslint-plugin-vuejs-accessibility`, but then I accidentally removed `eslint-plugin-vue` in resolving the merge conflict, and then I tried to fix it but I put the wrong version, and then I finally pushed it with the right version. TL;DR: silly mistakes

peepopoggers commented:
@jasonhenriquez Is the package-json.lock change needed, because IMO package updates should be seperate from this pr.

jasonhenriquez commented:
> @jasonhenriquez Is the package-json.lock change needed / possible to remove commit, because IMO package updates should be seperate from this pr.

See discussion under #1580:
> #1525 now successfully adds eslint-plugin-vuejs-accessibility and implements its recommended linter rules. It also completely satisfies the new rules (as of 0e1f700). I know it is irksome to link one PR to two issues, but this seems to be the most convenient means to have the full project pass these accessibility rules. @PrestonN @GilgusMaximus Please let me know if this is acceptable.

peepopoggers commented:
@jasonhenriquez Alright +1

ChunkyProgrammer commented on src/renderer/components/ft-select/ft-select.vue:
```diff
@@ -19,9 +19,9 @@
     <span class="select-highlight" />
     <span class="select-bar" />
-    <label class="select-label">
+    <p class="select-label">
```
Should clicking on one of the labels focus the select?

jasonhenriquez replied:
Can you clarify the question?

ChunkyProgrammer replied:
I'm just curious why you switched to paragraph tags instead of adding a "for" attribute to the label. 

I don't think the paragraphs are binding properly with the form elements (when you click on a label it normally focuses the associated input element & this doesn't happen with the paragraphs).

This link describes the importance of labels for accessibility:
https://developer.mozilla.org/en-US/docs/Learn/Forms/How_to_structure_a_web_form#the_<label>_element

This link has good alternatives to label (ex. Adding `v-bind:aria-labelby="placeholder"` to the select elements ):
https://www.w3.org/WAI/tutorials/forms/labels/

jasonhenriquez replied:
You're completely correct, fixed now. I thought that for some reason our implementation did not support this behavior. I tried again after you posted this and fiddled around for an hour trying to figure out why it was not working. It turns out that the label was given the `pointer-events: none` style in `ft-select.css`.

ChunkyProgrammer commented on src/renderer/components/ft-select/ft-select.vue:
```diff
@@ -1,8 +1,10 @@
 <template>
   <div class="select">
     <select
+      :id="placeholder"
```
This solution works well. Great job! One minor change that I'd suggest is doing a regex on the placeholder for ":id"/":for" to remove the whitespace as html ids shouldn't contain whitespace ex: `placeholder.replace(/ /g, '') `

jasonhenriquez replied:
This is something I did not know! It is done.

RastislavKish commented:
@jasonhenriquez I was discussing this with other blind fellows, and there are two additional things that could be done to improve things even further.
* On the page with opened video, the video title heading is currently not a heading, just text. Similarly, when opening a channel's site, the title of the channel as I understand has the role of heading, but is not marked like one.
* When activating subtitles, screenreaders currently don't speak them aloud if they appear. May be marking them live regions could resolve this?

Do you think it would be possible to incorporate these?
It's a very pleasant experience already, would make things perfect!

Thanks!

jasonhenriquez commented:
@RastislavKish Thank you for bringing this up! I have just pushed two commits to address your two bullet points. I am confident that I addressed your first bullet point, but please let me know if subtitles are working now. My suspicion is that content with the `pointer-events: none` style are ignored by many screen readers, and I have just removed that style attribute from the subtitles.

RastislavKish commented:
@jasonhenriquez Yes, the firstone is resolved, thanks!

The subtitles are not working, I've tried on both Linux and Windows.
I can see they're being shown if I manually navigate around, but they're not read automatically.

Does their paragraph, or whatever element it is, have aria-live attribute set at least to polite?
Is it one persistent element, or does a newone appear every time a new subtitle is displayed?

jasonhenriquez commented:
> @jasonhenriquez Yes, the firstone is resolved, thanks!
> 
> The subtitles are not working, I've tried on both Linux and Windows.
> I can see they're being shown if I manually navigate around, but they're not read automatically.
> 
> Does their paragraph, or whatever element it is, have aria-live attribute set at least to polite?
> Is it one persistent element, or does a newone appear every time a new subtitle is displayed?

We use the 'subtitles' feature of video.js to generate the captions. Video.js subtitles are set to have aria-live set to false and aria-atomic set to true by design, and I am having difficulty changing this manually. [The rationale for this implementation is talked about here.](https://github.com/videojs/video.js/issues/3554) Do you disagree with their explanation, and/or is there another way that you would prefer this behavior implemented?

RastislavKish commented:
@jasonhenriquez ah I see, then it'd probably be better to open an issue on the player's GitHub, since it's a general thing not specific to FreeTube.

Honestly, their rationale doesn't quite make a sense to me. Like yes, subtitles can be used in case of hearing impairment, but they're also widely used for translations, either of just parts of or entire videos.
As long as they're off by default and activated only when the user wants them, automatic reading should be the primary behavior.

I will discuss this with our community if we're of the same mindset and bring it up on the appropriate place, thanks for referring!

The manual reading has worked before as well, the original styling should be okay I guess.

jasonhenriquez commented:
@RastislavKish Glad to be of help! I agree, there should at the very least be an option (e.g., through an additional argument) to set aria-live to polite for captions and subtitles. In the unlikely worst case scenario that the Video.js team does not welcome your suggestion, I will gladly try to rig up a fix for this problem.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
@jasonhenriquez After reading ur last comment i think i gave this pr the wrong label by accident. Feel free to remove it if i mislabeled it.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Is it possible to fix this behaviour in this PR or should i create an issue for it.


https://user-images.githubusercontent.com/73130443/134637967-699b7ad0-5259-48be-b8ac-98dc568a3157.mp4

jasonhenriquez commented:
Should be a separate fix IMO. The fix would be hard coded to this particular case — app [re]load, on Subscriptions page, & attempted navigation to Subscriptions page. I have seen this before, and it does not occur even on reload when starting on another route. I suspect it has to do with the specialness of the Subscriptions page as the default starting page, just in case you need that info.

On Fri, Sep 24, 2021 at 02:49, efb4f5ff-1298-471a-8973-3d47447115dc ***@***.***> wrote:

> Is it possible to fix this behaviour in this PR or should i create an issue for it.
>
> https://user-images.githubusercontent.com/73130443/134637967-699b7ad0-5259-48be-b8ac-98dc568a3157.mp4
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/FreeTubeApp/FreeTube/pull/1525#issuecomment-926420628), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AUHXK2WNV23CHVFSTETADFTUDQUQBANCNFSM5AQAGJ2A).
> Triage notifications on the go with GitHub Mobile for [iOS](https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675) or [Android](https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub).

jasonhenriquez commented:
@RastislavKish Given that video.js has closed your issue due to inactivity, I am going to consider this blocking issue ongoing. We can create an issue for it here, but we will have to examine our options for solving that in a separate future effort. This PR has more than enough features as-is, and any additional accessibility issues identified should be posted in the FreeTube issues board as well. Thank you for all of the assistance and effort you provided in helping to review & propose changes. Because of your feedback & contributions, we are much further than we had ever previously endeavored on our path to making FreeTube a pleasant and accessible app for everyone.

This PR is no longer a work-in-progress and is ready for testing.

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (approved):
Didnt find anything weird LGTM

PikachuEXE commented on src/renderer/components/ft-icon-button/ft-icon-button.js:
```diff
@@ -114,7 +126,19 @@ export default Vue.extend({
+        return
+      }
+
+      // const listbox = $(`#${this.id}`)
+      // const allOptions = listbox.children()
+
+      // allOptions.attr('aria-selected', 'false')
+      // allOptions.attr('tabindex', '-1')
+      // event.target.setAttribute('aria-selected', 'true')
+      // event.target.setAttribute('tabindex', '0')
```
What happen to this commented code?

ChunkyProgrammer replied:
@jasonhenriquez

ChunkyProgrammer replied:
@jasonhenriquez would you be able to provide context on the removal? maybe someone else could work on it. (I think this is the main blocker for the code to be reviewed)

PikachuEXE reviewed (commented):
I looked around the code changes once
Too many changes so I might miss something
Not sure how to test locally

ChunkyProgrammer commented:
> I looked around the code changes once Too many changes so I might miss something Not sure how to test locally

Use this build as a daily driver and see if anything is broken

PikachuEXE commented:
@ChunkyProgrammer Cannot until 0.17 released...

ChunkyProgrammer commented on src/renderer/App.js:
```diff
@@ -541,3 +541,54 @@ export default Vue.extend({
 })
+
+// return true if it is a selection event, false if it is a navigation event
+Vue.prototype.$handleDropdownKeyboardEvent = function(event, target, afterElement) {
```
should probably convert these "Vue.prototypes" to mixins

ChunkyProgrammer commented on package.json:
```diff
@@ -42,6 +42,8 @@
     "rebuild:electron": "electron-builder install-app-deps",
     "rebuild:node": "npm rebuild",
     "release": "run-s test build",
+    "test": "run-s rebuild:node pack:workers jest",
```
```suggestion
```

ChunkyProgrammer commented on package.json:
```diff
@@ -42,6 +42,8 @@
     "rebuild:node": "npm rebuild",
     "release": "run-s test build",
+    "test": "run-s rebuild:node pack:workers jest",
+    "test:watch": "run-s rebuild:node pack:workers jest:watch",
```
```suggestion
```

ChunkyProgrammer commented on package.json:
```diff
@@ -104,6 +106,8 @@
     "eslint-plugin-standard": "^5.0.0",
     "eslint-plugin-vue": "^7.17.0",
+    "eslint-plugin-vuejs-accessibility": "^1.2.0",
+    "fast-glob": "^3.2.7",
```
a newer version of fast-glob is available. also curious as to why this package is added? is it needed by vuejs-accessibility?

ChunkyProgrammer replied:
```suggestion
```

ChunkyProgrammer commented on .eslintrc.js:
```diff
@@ -38,6 +42,12 @@ module.exports = {
     'no-undef': 1,
-    'vue/no-template-key': 1
+    'vue/no-template-key': 1,
+    'vuejs-accessibility/no-onchange': 0,
```
```suggestion
    'vuejs-accessibility/no-onchange': 'warn',
```

ChunkyProgrammer commented:
Going to keep this open for now

ChunkyProgrammer commented:
Most of the changes in this PR have been implemented.
Things that still need to be implemented:
- accessibility improvements to video/playlist/channel lists
- profile elements 

Also created a C# project for selenium tests (it runs accessibility analysis agaianst FreeTube Cordova) https://github.com/ChunkyProgrammer/FreeTube-Cordova-Selenium-Tests

efb4f5ff-1298-471a-8973-3d47447115dc commented:
@ChunkyProgrammer i wanted to test FT in more places just like i reported in https://github.com/FreeTubeApp/FreeTube/issues/3643

Should i keep opening issues or is it best to put  all my findings here?

ChunkyProgrammer commented:
From what I can tell, the only thing from this PR that still needs to be added to FreeTube is using headings (h tags) for search results

efb4f5ff-1298-471a-8973-3d47447115dc commented:
@ChunkyProgrammer do you think this can be closed now that your latest PR has been merged?
